### PR TITLE
feat: Support Optimism Fault Proofs

### PR DESCRIPTION
**Pull request by varasev**
Closes https://github.com/blockscout/blockscout/issues/9893.

## Motivation

This PR adds support of Optimism Fault Proofs to correctly determine statuses of withdrawals indexed by `Indexer.Fetcher.Optimism.Withdrawal` module.

Also, the new `Indexer.Fetcher.Optimism.DisputeGame` module is introduced that indexes dispute games used by the Withdrawal module to get withdrawal status.

The changes are backward compatible with pre-Fault Proofs implementation and activated seamlessly after `OptimismPortal` contract upgrades.

Now, the withdrawal status can be one of:
- `Waiting for state root` (when waiting for appropriate dispute games to appear in `DisputeGameFactory` contract)
- `Ready to prove` (when appropriate dispute games found in `DisputeGameFactory` contract)
- `Waiting a game to resolve` (when `WithdrawalProven` is emitted on L1 by `OptimismPortal` contract but the dispute game bound to the withdrawal is not resolved to `Defender wins` status yet)
- `In challenge period` (when the withdrawal must wait until it's allowed to be finalized)
- `Ready for relay` (when the withdrawal is ready to be finalized)
- `Proven` (replaces the above three statuses in case of the module cannot determine a game bound to the withdrawal)
- `Relayed` (when `WithdrawalFinalized` event is emitted on L1 by `OptimismPortal` contract and the withdrawal is finalized)

To see the games indexed, a new API endpoint was added: `/api/v2/optimism/games` (and `/api/v2/optimism/games/count`) with the following fields for each JSON item:
- `index` (numeric)
- `game_type` (numeric)
- `contract_address` (string in the form of `0x...`)
- `l2_block_number` (numeric)
- `created_at` (datetime, e.g. `2024-04-16T09:56:48.000000Z`)
- `status` (string, can be one of: `In progress`, `Challenger wins`, `Defender wins`)
- `resolved_at` (null or datetime)

Example of the API output (truncated):

```
{"items":[{"contract_address":"0xa1f95f4387bb74050df37ef0a5672bdc7ef9ffd1","created_at":"2024-04-16T09:56:48.000000Z","game_type":0,"index":2647,"l2_block_number":10729366,"resolved_at":null,"status":"In progress"},...,{"contract_address":"0xb128c980db9d8f9838d0c639e8b5e094cf112fbd","created_at":"2024-04-15T21:41:48.000000Z","game_type":0,"index":2598,"l2_block_number":10707321,"resolved_at":null,"status":"In progress"}],"next_page_params":{"index":2598,"items_count":50}}
```

## Checklist for your Pull Request (PR)

  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [ ] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I added new DB indices, I checked, that they are not redundant with PGHero or other tools.
  - [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

### CODE REVIEW

File: apps/indexer/lib/indexer/fetcher/optimism/dispute_game.ex [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+      end)
+
+    error_message = &"Cannot call gameAtIndex() public getter of DisputeGameFactory. Error: #{inspect(&1)}"
+
+    with {:ok, responses} <-
+           IndexerHelper.repeated_call(&json_rpc/2, [requests, json_rpc_named_arguments], error_message, retries),
+         games = decode_games(responses),
+         extra_data_by_index = read_extra_data("0x609d3334", "extraData()", games, json_rpc_named_arguments),
+         false <- is_nil(extra_data_by_index),
+         resolved_at_by_index = read_extra_data("0x19effeb4", "resolvedAt()", games, json_rpc_named_arguments),
```
Comment (vbaranov) [inline apps/indexer/lib/indexer/fetcher/optimism/dispute_game.ex:343]: "0x19effeb4" and "0x200d2ed2" are defined twice in the module. Let's extract to module attribute.

File: apps/explorer/lib/explorer/chain/import/runner/optimism/dispute_games.ex [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  @impl Import.Runner
+  def timeout, do: @timeout
+
+  @spec insert(Repo.t(), [map()], %{required(:timeout) => timeout(), required(:timestamps) => Import.timestamps()}) ::
+          {:ok, [DisputeGame.t()]}
+          | {:error, [Changeset.t()]}
+  def insert(repo, changes_list, %{timeout: timeout, timestamps: timestamps} = options) when is_list(changes_list) do
+    on_conflict = Map.get_lazy(options, :on_conflict, &default_on_conflict/0)
+
+    # Enforce DisputeGame ShareLocks order (see docs: sharelock.md)
```
[the remark this answers is not in the document]
Comment (varasev) [reply apps/explorer/lib/explorer/chain/import/runner/optimism/dispute_games.ex:64]: Yes, done in https://github.com/blockscout/docs/pull/265

File: apps/explorer/lib/explorer/chain/optimism/withdrawal.ex [hunk 3]
```diff
@@ -113,51 +124,170 @@ defmodule Explorer.Chain.Optimism.Withdrawal do
...
 
-  def status(_w) do
-    {"Relayed", nil}
+  defp post_fault_proofs_status(l1_timestamp, game) do
+    if game.status != @game_status_defender_wins do
+      # the game status is not DEFENDER_WINS
+      {@withdrawal_status_waiting_to_resolve, nil}
+    else
+      dispute_game_finality_delay_seconds =
+        Helper.parse_integer(Constants.get_constant_value("optimism_dispute_game_finality_delay_seconds"))
```
[the remark this answers is not in the document]
Comment (varasev) [reply apps/explorer/lib/explorer/chain/optimism/withdrawal.ex]: Defined in https://github.com/blockscout/blockscout/pull/9892/commits/dfdfa6c23c51b3f541ad990a418b00525e5e1e85

File: apps/indexer/lib/indexer/fetcher/optimism/withdrawal_event.ex [hunk 4]
```diff
@@ -217,4 +255,18 @@ defmodule Indexer.Fetcher.Optimism.WithdrawalEvent do
...
   end
+
+  defp input_to_game_index(input) do
+    method_signature = String.slice(input, 0..9)
+
+    if method_signature == "0x4870496f" do
+      # the signature of `proveWithdrawalTransaction(tuple _tx, uint256 _disputeGameIndex, tuple _outputRootProof, bytes[] _withdrawalProof)` method
+      {game_index, ""} =
+        input
+        |> String.slice(74..137)
```
[the remark this answers is not in the document]
Comment (varasev) [reply apps/indexer/lib/indexer/fetcher/optimism/withdrawal_event.ex]: I clarified how to calculate the range in https://github.com/blockscout/blockscout/pull/9892/commits/dfdfa6c23c51b3f541ad990a418b00525e5e1e85

File: apps/indexer/lib/indexer/fetcher/optimism/dispute_game.ex (hunk 1 above)
Comment (varasev) [reply apps/indexer/lib/indexer/fetcher/optimism/dispute_game.ex]: Extracted in https://github.com/blockscout/blockscout/pull/9892/commits/dfdfa6c23c51b3f541ad990a418b00525e5e1e85

**Review by vbaranov — APPROVED** [review]

**Pull request merged**